hazelcast-client/src/main/java/com/hazelcast/client/util/ClientDelegatingFuture.java

```diff
@@ -189,86 +142,41 @@ public V join() {
         }
     }
 
-    private Object getResult() {
+    private Object resolveResponse(ClientMessage clientMessage, boolean deserialize) {
         if (defaultValue != null) {
             return defaultValue;
         }
 
-        // If value is already deserialized, use it.
-        if (deserializedValue != null) {
-            return deserializedValue;
+        ClientMessage tempMessage = ClientMessage.createForDecode(clientMessage.buffer(), 0);
+        Object decodedResponse = clientMessageDecoder.decodeClientMessage(tempMessage);
```

Comment:
`clientMessageDecoder.decodeClientMessage()` creates a copy of internal buffer each time. 

@pveentjer: Which one do you prefer; 
- allocating memory and creating temp Data object on each invocation 
or 
- guarding the method with a lock and creating a single Data object only?